### DISCUSSION TITLE
Renaming/topology issue

### DISCUSSION THREAD

**Discussion by raharth** [FreeCAD/FreeCAD#10822] [Discussions that will turn in to issues but need refining]
I just started using FreeCAD and I immediately stumbled into the topology issue with several parts being thrown around all over the place, when altering previously made sketches.  I did some research on the topic and as far as I understand one of the issues causing this is that elements are renamed with the constraints not properly updated? At least that's the only way that would explain how elements are move accross the entire design attached to edges that are not anyhow involved in that part originally.
I'm a mathematician and I solve algorithmic problems for living so I'm happy to have a look at it if someone could give me some background on the issue and why it has not been fixed yet (I guess some hidden stuff deep down in the code?).
I'm specifically talking about the issue shown in this video:
https://www.youtube.com/watch?v=GKn5kAYmfQE&t=612s

Two questions I'd have assuming that indeed the renaming is the issue in combination with faulty updates on the constraints:
1. Why does it need to be renamed in the first place?
2. If it needs to be renamed, is there a reason to not do it in a two step process?
3. Can someone point me to where this naming update is happening in the code?

Thank you very much! :)

**Comment by chennes** [comment]
For the solution we are currently in the process of merging, see the writeups:
https://github.com/realthunder/FreeCAD_assembly3/wiki/Topological-Naming
and
https://github.com/realthunder/FreeCAD_assembly3/wiki/Topological-Naming-Algorithm

Those documents will give you some background on the problem, some academic research into its mitigation, and quite a bit of detail about how we are going about implementing that.

**Reply by raharth** [reply]
Great, thank you!

Edit: So you consider this issue as currently solved but not yet merged?

**Reply by chennes** [reply]
"Solved" is not the term we use: let's go with "mitigated" 😄.

**Comment by raharth** [comment]
@chennes Hi Chris, is there any time line at what point you have planed to merge this fix? Removing a part of a sketch has just once again completely messed with my work. I think FreeCAD really has a lot of potential but I spent a multiple of time trying to fix issues caused by the renaming and topology, so for know I will have to move away from it. It's just unusable for me. It's really sad and I hope you get that issue fixed, if you do I'm certainly back to using it in no time, but yeah... also reading things like "that's known behavior hence no issue" in some of the tickets on that issue is not particularly reassuring. I'm not trying to bash you or the project as I said I think it has a lot of potential, but not as long this remains an open issue - basically I just want to raise awareness on that, maybe it helps maybe not. Feel free to ignore it entirely. That being said I hope to be back soon.

Best

**Reply by chennes** [reply]
Yes: we expect the full merge to take something on the order of six months (from now, not from when we started). Meaningful improvement will occur in a month or so, but come with significant performance penalties. We expect to address those shortly thereafter.

**Reply by raharth** [reply]
I'm 100% fine with the performance suffering as long as my designes stop breaking all the time with any changes! :D 
Looking forward to those changes!
